pkg/stargate/deployments.go

```diff
@@ -0,0 +1,324 @@
+package stargate
+
+import (
+	"fmt"
+	cassdcapi "github.com/k8ssandra/cass-operator/operator/pkg/apis/cassandra/v1beta1"
+	api "github.com/k8ssandra/k8ssandra-operator/api/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/cassandra"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"math"
+	"strings"
+)
+
+const (
+	cassandraConfigDir  = "/config"
+	cassandraConfigPath = "/config/cassandra.yaml"
+
+	// FIXME should this be customized? Cf. K8ssandra 1.x Helm chart template:
+	// "{{ .Values.clusterDomain | default \"cluster.local\" }}
+	clusterDomain = "cluster.local"
+)
+
+// NewDeployments compute the Deployments to create for the given Stargate and CassandraDatacenter
+// resources.
+func NewDeployments(stargate *api.Stargate, dc *cassdcapi.CassandraDatacenter) map[string]appsv1.Deployment {
+
+	clusterVersion := computeClusterVersion(dc)
+	seedService := computeSeedServiceUrl(dc)
+
+	racks := dc.GetRacks()
+
+	averageReplicasPerRack := int32(math.Ceil(float64(stargate.Spec.Size) / float64(len(racks))))
+	replicasRemaining := stargate.Spec.Size
+
+	var deployments = make(map[string]appsv1.Deployment)
+	for _, rack := range racks {
+
+		if replicasRemaining == 0 {
+			// No more room for more Stargate pods, some racks won't have any Stargate pod
```

Comment:
I feel like we should do something here to let the user know that some racks won't have Stargate pods. Putting something in the operator logs is fine but it doesn't have good visibility. 

There is a broader question that we should consider which is - should we even allow uneven racks?

Reply:
I will add a warning for now, not sure about what we could do more.

> should we even allow uneven racks?

I checked the cass-operator code and it does seem to allow uneven racks, see function `cassdcapi.SplitRacks`. I was also able to execute the following test:

```
func TestUnevenRacks(t *testing.T) {
	rc, _, _ := setupTest()
	rc.Datacenter.Spec.Racks = []api.Rack{{Name: "rack0"}, {Name: "rack1"}, {Name: "rack2"}}
	rc.Datacenter.Spec.Size = 4
	err := rc.CalculateRackInformation()
	assert.NoError(t, err)
	assert.Equal(t, 2, rc.desiredRackInformation[0].NodeCount)
	assert.Equal(t, 1, rc.desiredRackInformation[1].NodeCount)
	assert.Equal(t, 1, rc.desiredRackInformation[2].NodeCount)
}
```

So my suggestion would be that uneven rack should be allowed in Stargate as well.

Reply:
IIRC I thought we blocked uneven racks in the validating webhook

Reply:
I don't see anything in cass-operator webhooks preventing uneven racks, but maybe I'm not looking at the right place. In any case, I'd vote for allowing uneven racks in Stargate deployments: it doesn't change much our code and the impacts of uneven Stargate pod distributions seems low to me. I added a message in the logs already if that happens.